broker/fragment/store_azure.go

```diff
@@ -222,47 +237,133 @@ func getAzureStorageCredential(coreCredential azcore.TokenCredential, tenant str
 	return credential, nil
 }
 
-func (a *azureBackend) azureClient(ep *url.URL) (cfg AzureStoreConfig, client pipeline.Pipeline, err error) {
-	if err = parseStoreArgs(ep, &cfg); err != nil {
+func parseAzureEndpoint(endpoint *url.URL) (cfg azureStoreConfig, err error) {
+	if err = parseStoreArgs(endpoint, &cfg); err != nil {
 		return
 	}
+
 	// Omit leading slash from URI. Note that FragmentStore already
 	// enforces that URL Paths end in '/'.
-	var splitPath = strings.Split(ep.Path[1:], "/")
+	var splitPath = strings.Split(endpoint.Path[1:], "/")
 
-	if ep.Scheme == "azure" {
+	if endpoint.Scheme == "azure" {
 		// Since only one non-ad "Shared Key" credential can be injected via
 		// environment variables, we should only keep around one client for
 		// all `azure://` requests.
 		cfg.accountTenantID = "AZURE_SHARED_KEY"
 		cfg.storageAccountName = os.Getenv("AZURE_ACCOUNT_NAME")
-		cfg.containerName, cfg.prefix = ep.Host, ep.Path[1:]
-	} else if ep.Scheme == "azure-ad" {
-		cfg.accountTenantID, cfg.storageAccountName, cfg.containerName, cfg.prefix = ep.Host, splitPath[0], splitPath[1], strings.Join(splitPath[2:], "/")
+		cfg.containerName, cfg.prefix = endpoint.Host, endpoint.Path[1:]
+	} else if endpoint.Scheme == "azure-ad" {
+		cfg.accountTenantID, cfg.storageAccountName, cfg.containerName, cfg.prefix = endpoint.Host, splitPath[0], splitPath[1], strings.Join(splitPath[2:], "/")
 	}
 
-	a.clientMu.Lock()
-	defer a.clientMu.Unlock()
+	return cfg, nil
+}
 
-	if a.clients[cfg.accountTenantID] != nil {
-		client = a.clients[cfg.accountTenantID]
-		return
+func (a *azureBackend) getAzureServiceClient(endpoint *url.URL) (client *service.Client, err error) {
+	var cfg azureStoreConfig
+
+	if cfg, err = parseAzureEndpoint(endpoint); err != nil {
+		return nil, err
 	}
 
-	var credentials azblob.Credential
+	a.mu.Lock()
+	if a.svcClients == nil {
+		a.svcClients = make(map[string]service.Client)
+	}
+	a.mu.Unlock()
 
-	if ep.Scheme == "azure" {
+	if endpoint.Scheme == "azure" {
 		var accountName = os.Getenv("AZURE_ACCOUNT_NAME")
 		var accountKey = os.Getenv("AZURE_ACCOUNT_KEY")
+
+		a.mu.Lock()
+		client, ok := a.svcClients[accountName]
+		a.mu.Unlock()
+
+		if ok {
+			log.WithFields(log.Fields{
+				"storageAccountName": accountName,
+			}).Info("Re-using cached azure:// service client")
+			return &client, nil
+		}
+
 		sharedKeyCred, err := service.NewSharedKeyCredential(accountName, accountKey)
 		if err != nil {
-			return cfg, nil, err
+			return nil, err
 		}
 		serviceClient, err := service.NewClientWithSharedKeyCredential(cfg.serviceUrl(), sharedKeyCred, &service.ClientOptions{})
 		if err != nil {
-			return cfg, nil, err
+			return nil, err
 		}
-		a.svcClient = *serviceClient
+
+		a.mu.Lock()
+		a.svcClients[accountName] = *serviceClient
```

Comment:
You're dereferencing (making a shallow copy) here and I don't understand why. Seems like `svcClients` should be holding pointers, since that's what's returned by service.NewClientWithSharedKeyCredential().

Reply:
Fixed, good point about `svcClients` (now just `clients`) holding a map of pointers.